File: 3534-path-existence-queries-in-a-graph-ii/3534-path-existence-queries-in-a-graph-ii.py
class Solution:
    def pathExistenceQueries(self, n: int, nums: List[int], maxDiff: int, queries: List[List[int]]) -> List[int]:
        sorted_pairs = [(nums[i], i) for i in range(n)]
        sorted_pairs.sort(key=lambda x: x[0])

        index_map = {idx: i for i, (_, idx) in enumerate(sorted_pairs)}

        LOG = 1
        while (1 << LOG) < n:
            LOG += 1

        up = [[0] * LOG for _ in range(n)]
        r = 0
        for l in range(n):
            if r < l:
                r = l
            while r + 1 < n and sorted_pairs[r + 1][0] - sorted_pairs[l][0] <= maxDiff:
                r += 1
            up[l][0] = r

        for j in range(1, LOG):
            for ind in range(n):
                up[ind][j] = up[up[ind][j - 1]][j - 1]

        ans = [0] * len(queries)
        for ind, (src_orig, dest_orig) in enumerate(queries):
            src = index_map[src_orig]
            dest = index_map[dest_orig]

            if src > dest:
                src, dest = dest, src

            if src == dest:
                ans[ind] = 0
                continue

            curr = src
            hops = 0
            for j in range(LOG - 1, -1, -1):
                if up[curr][j] < dest:
                    curr = up[curr][j]
                    hops += (1 << j)

            if up[curr][0] >= dest:
                ans[ind] = hops + 1
            else:
                ans[ind] = -1

        return ans

File: 3534-path-existence-queries-in-a-graph-ii/3534-path-existence-queries-in-a-graph-ii.cpp
class Solution {
public:
    vector<int> pathExistenceQueries(int n, vector<int>& nums, int maxDiff, vector<vector<int>>& queries) {
        vector<pair<int,int>> sorted(n);
        for(int i = 0; i < n; i++) {
            sorted[i] = {nums[i], i};
        }

        sort(sorted.begin(), sorted.end(), [](const pair<int,int>& a, const pair<int,int>& b) {
            return a.first < b.first;
        });

        unordered_map<int,int> map;
        for(int i = 0; i < n; i++) {
            map[sorted[i].second] = i;
        }
        int LOG = 1;
        while ((1 << LOG) < n) LOG++;
        vector<vector<int>> up(n, vector<int>(LOG));
        int r = 0;
        for (int l = 0; l < n; l++) {
            if (r < l) r = l;                      
            while (r + 1 < n && sorted[r + 1].first - sorted[l].first <= maxDiff) r++;
            up[l][0] = r;                         
        }

        for (int j = 1; j < LOG; j++) {
            for (int ind = 0; ind < n; ind++) {
                up[ind][j] = up[up[ind][j - 1]][j - 1];
            }
        }

        vector<int> ans(queries.size());
        for(size_t ind = 0; ind < queries.size(); ind++) {
            int src = map[queries[ind][0]];
            int dest = map[queries[ind][1]];

            if(src > dest) {
                swap(src, dest); 
            }

            if(src == dest) {
                ans[ind] = 0; 
                continue;
            }

            int curr = src;
            int hops = 0;
            for (int j = LOG - 1; j >= 0; j--) {
                if (up[curr][j] < dest) {
                    curr = up[curr][j];
                    hops += (1 << j);
                }
            }
            if (up[curr][0] >= dest) {
                ans[ind] = hops + 1;
            } else {
                ans[ind] = -1; 
            }
        }

        return ans;
    }
};